File: Source/Loaders/Android-IME/app/src/main/java/tw/chichi77/keykey/android/SmartMandarinStore.java
package tw.chichi77.keykey.android;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Lazy, read-only Viterbi walker over the same language model used by macOS and iOS. */
final class SmartMandarinStore implements SmartMandarinSource, AutoCloseable {
    private static final String ASSET_NAME = "KeyKey.db";
    private static final String INSTALLED_NAME = "KeyKey-smart-reading-v3.db";
    private static final String[] PREVIOUS_INSTALLED_NAMES = {
            "KeyKey-smart-885614.db", "KeyKey-smart-1.2.10.db",
            "KeyKey-smart-reading-v2.db"
    };
    private static final long EXPECTED_BIGRAM_ROWS = 885_627;
    private static final int MAXIMUM_SPAN = 8;

    private record Unigram(String text, double probability, double backoff) {}
    private record BigramKey(String previous, String current) {}
    private record Path(double score, double lastBackoff,
                        List<SmartMandarinSegment> segments) {}

    private final SQLiteDatabase database;
    private final SmartMandarinUserData userData;
    private final Map<String, List<Unigram>> unigramCache = new HashMap<>();
    private final Map<String, Map<BigramKey, Double>> bigramCache = new HashMap<>();

    static SmartMandarinStore open(Context context, SmartMandarinUserData userData)
            throws IOException {
        File databaseFile = new File(context.getNoBackupFilesDir(), INSTALLED_NAME);
        if (!databaseFile.isFile() || databaseFile.length() == 0) {
            File temporary = new File(databaseFile.getParentFile(), INSTALLED_NAME + ".tmp");
            try (InputStream input = context.getAssets().open(ASSET_NAME);
                 FileOutputStream output = new FileOutputStream(temporary)) {
                byte[] buffer = new byte[64 * 1024];
                int count;
                while ((count = input.read(buffer)) >= 0) output.write(buffer, 0, count);
                output.getFD().sync();
            } catch (IOException error) {
                temporary.delete();
                throw error;
            }
            if (!temporary.renameTo(databaseFile)) {
                temporary.delete();
                throw new IOException("Unable to install " + ASSET_NAME);
            }
        }
        SQLiteDatabase database = SQLiteDatabase.openDatabase(
                databaseFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READONLY);
        try (Cursor cursor = database.rawQuery("SELECT COUNT(*) FROM bigrams", null)) {
            if (!cursor.moveToFirst() || cursor.getLong(0) != EXPECTED_BIGRAM_ROWS) {
                throw new IOException("The bundled Smart Mandarin database must have 885627 bigrams");
            }
        } catch (IOException | RuntimeException error) {
            database.close();
            throw error;
        }
        for (String previousName : PREVIOUS_INSTALLED_NAMES) {
            new File(databaseFile.getParentFile(), previousName).delete();
        }
        return new SmartMandarinStore(database, userData);
    }

    SmartMandarinStore(SQLiteDatabase database) {
        this(database, null);
    }

    SmartMandarinStore(SQLiteDatabase database, SmartMandarinUserData userData) {
        this.database = database;
        this.userData = userData;
    }

    @Override
    public SmartMandarinComposition compose(List<String> readings,
                                             Map<Integer, String> overrides) {
        Map<Integer, SmartMandarinSelection> selections = new HashMap<>();
        for (Map.Entry<Integer, String> entry : overrides.entrySet()) {
            selections.put(entry.getKey(), new SmartMandarinSelection(1, entry.getValue()));
        }
        return composeSelections(readings, selections);
    }

    @Override
    public SmartMandarinComposition composeSelections(
            List<String> readings, Map<Integer, SmartMandarinSelection> selections) {
        if (readings.isEmpty()) return new SmartMandarinComposition("", List.of());

        List<Map<String, Path>> paths = new ArrayList<>(readings.size() + 1);
        for (int index = 0; index <= readings.size(); index++) paths.add(new HashMap<>());
        paths.get(0).put("", new Path(0, 0, List.of()));

        for (int start = 0; start < readings.size(); start++) {
            if (paths.get(start).isEmpty()) continue;
            int largestSpan = Math.min(MAXIMUM_SPAN, readings.size() - start);
            for (int length = 1; length <= largestSpan; length++) {
                int end = start + length;
                String query = String.join("", readings.subList(start, end));
                int overlapCount = 0;
                for (Map.Entry<Integer, SmartMandarinSelection> selection : selections.entrySet()) {
                    int selectedStart = selection.getKey();
                    if (selectedStart < end && start < selectedStart + selection.getValue().length()) {
                        overlapCount++;
                    }
                }
                SmartMandarinSelection required = selections.get(start);
                if (overlapCount != 0
                        && !(overlapCount == 1 && required != null
                        && required.length() == length)) continue;

                List<Unigram> entries = unigrams(query);
                String learned = learnedCandidate(query);
                if (required != null) {
                    ArrayList<Unigram> matching = new ArrayList<>();
                    for (Unigram row : entries) {
                        if (row.text().equals(required.text())) matching.add(row);
                    }
                    entries = matching;
                }
                if (entries.isEmpty()) continue;

                for (Unigram entry : entries) {
                    SmartMandarinSegment segment = new SmartMandarinSegment(
                            start, length, query, entry.text());
                    for (Path previousPath : paths.get(start).values()) {
                        Double bigram;
                        if (previousPath.segments().isEmpty()) {
                            bigram = bigramProbability("!", query, "", entry.text());
                        } else {
                            SmartMandarinSegment previous = previousPath.segments()
                                    .get(previousPath.segments().size() - 1);
                            bigram = bigramProbability(previous.query(), query,
                                    previous.text(), entry.text());
                        }
                        double fallback = previousPath.lastBackoff() + entry.probability();
                        double transition = bigram == null
                                ? fallback : Math.max(bigram, fallback);
                        ArrayList<SmartMandarinSegment> segments =
                                new ArrayList<>(previousPath.segments());
                        segments.add(segment);
                        if (entry.text().equals(learned)) {
                            transition = Math.max(transition, 0);
                        }
                        Path path = new Path(previousPath.score() + transition,
                                entry.backoff(), List.copyOf(segments));
                        String stateKey = query + '\u001f' + entry.text();
                        Path existing = paths.get(end).get(stateKey);
                        if (existing == null || path.score() > existing.score()) {
                            paths.get(end).put(stateKey, path);
                        }
                    }
                }
            }
        }

        Path best = null;
        double bestScore = Double.NEGATIVE_INFINITY;
        for (Path path : paths.get(readings.size()).values()) {
            // The desktop walker does not let the end marker replace the
            // first candidate of an isolated syllable.
            double score = readings.size() == 1 ? path.score() : finalScore(path);
            if (score > bestScore) {
                best = path;
                bestScore = score;
            }
        }
        if (best == null) return null;
        StringBuilder text = new StringBuilder();
        for (SmartMandarinSegment segment : best.segments()) text.append(segment.text());
        return new SmartMandarinComposition(text.toString(), best.segments());
    }

    @Override
    public List<String> candidates(List<String> readings, int index,
                                   SmartMandarinComposition composition) {
        ArrayList<String> texts = new ArrayList<>();
        for (SmartMandarinCandidate candidate : candidateOptions(readings, index, composition)) {
            texts.add(candidate.text());
        }
        return List.copyOf(texts);
    }

    @Override
    public List<SmartMandarinCandidate> candidateOptions(
            List<String> readings, int index, SmartMandarinComposition composition) {
        if (index < 0 || index >= readings.size()) return List.of();
        SmartMandarinSegment previous = null;
        if (composition != null) {
            for (SmartMandarinSegment segment : composition.segments()) {
                if (segment.start() + segment.length() == index) previous = segment;
            }
        }

        record Ranked(SmartMandarinCandidate candidate, double score) {}
        ArrayList<Ranked> ranked = new ArrayList<>();
        double previousBackoff = 0;
        if (previous != null) {
            for (Unigram item : unigrams(previous.query())) {
                if (item.text().equals(previous.text())) {
                    previousBackoff = item.backoff();
                    break;
                }
            }
        }
        int largestSpan = Math.min(MAXIMUM_SPAN, readings.size() - index);
        for (int length = 1; length <= largestSpan; length++) {
            String query = String.join("", readings.subList(index, index + length));
            String learned = learnedCandidate(query);
            for (Unigram entry : unigrams(query)) {
                Double bigram = previous == null
                        ? bigramProbability("!", query, "", entry.text())
                        : bigramProbability(previous.query(), query,
                                previous.text(), entry.text());
                double fallback = previousBackoff + entry.probability();
                ranked.add(new Ranked(new SmartMandarinCandidate(length, entry.text()),
                        entry.text().equals(learned) ? 0
                                : bigram == null ? fallback : Math.max(bigram, fallback)));
            }
        }
        ranked.sort((left, right) -> Double.compare(right.score(), left.score()));
        LinkedHashSet<SmartMandarinCandidate> unique = new LinkedHashSet<>();
        for (Ranked item : ranked) unique.add(item.candidate());
        return List.copyOf(unique);
    }

    @Override
    public void learnSelection(List<String> readings, int index, String selected,
                               SmartMandarinComposition composition) {
        learnSelection(readings, index, new SmartMandarinCandidate(1, selected), composition);
    }

    @Override
    public void learnSelection(List<String> readings, int index,
                               SmartMandarinCandidate candidate,
                               SmartMandarinComposition composition) {
        if (userData == null || index < 0 || candidate.length() <= 0
                || index + candidate.length() > readings.size()) return;
        SmartMandarinSegment previous = null;
        if (composition != null) {
            for (SmartMandarinSegment segment : composition.segments()) {
                if (segment.start() + segment.length() == index) previous = segment;
            }
        }
        try {
            userData.learnCandidate(String.join("", readings.subList(index,
                            index + candidate.length())), candidate.text(),
                    previous == null ? null : previous.query(),
                    previous == null ? null : previous.text());
        } catch (RuntimeException ignored) {
            // A temporarily unavailable user database must not stop text input.
        }
    }

    @Override
    public void learnConfirmedComposition(SmartMandarinComposition composition) {
        if (userData == null) return;
        try {
            userData.learnComposition(composition);
        } catch (RuntimeException ignored) {
            // The cooked model can still commit the completed text.
        }
    }

    @Override
    public int evictionLength(List<String> readings, SmartMandarinComposition composition) {
        if (composition.segments().isEmpty()) return 0;
        int firstLength = composition.segments().get(0).length();
        if (firstLength != 1) return firstLength;

        // A learned single character can split a visible dictionary phrase.
        // Keep that phrase together when it leaves the editable window.
        int length = 0;
        StringBuilder query = new StringBuilder();
        StringBuilder text = new StringBuilder();
        for (SmartMandarinSegment segment : composition.segments()) {
            if (segment.start() != length || length + segment.length() > readings.size()) break;
            for (int index = 0; index < segment.length(); index++) {
                query.append(readings.get(length + index));
            }
            length += segment.length();
            text.append(segment.text());
            if (length < 2) continue;
            if (length > MAXIMUM_SPAN) break;
            for (Unigram entry : unigrams(query.toString())) {
                if (entry.text().equals(text.toString())) return length;
            }
        }
        return firstLength;
    }

    private double finalScore(Path path) {
        if (path.segments().isEmpty()) return path.score();
        SmartMandarinSegment last = path.segments().get(path.segments().size() - 1);
        Double ending = bigramProbability(last.query(), "$", last.text(), "");
        return path.score() + (ending == null
                ? path.lastBackoff() : Math.max(ending, path.lastBackoff()));
    }

    private List<Unigram> unigrams(String query) {
        List<Unigram> cached = unigramCache.get(query);
        if (cached == null) {
            ArrayList<Unigram> rows = new ArrayList<>();
            try (Cursor cursor = database.rawQuery(
                "SELECT current, probability, backoff FROM unigrams "
                        + "WHERE qstring = ? ORDER BY probability DESC LIMIT 64",
                new String[]{query})) {
                while (cursor.moveToNext()) {
                    String text = cursor.getString(0);
                    if (text != null && !text.isEmpty()) {
                        rows.add(new Unigram(text, cursor.getDouble(1), cursor.getDouble(2)));
                    }
                }
            }
            cached = List.copyOf(rows);
            unigramCache.put(query, cached);
        }
        if (userData == null) return cached;
        HashMap<String, Unigram> merged = new HashMap<>();
        for (Unigram item : cached) merged.put(item.text(), item);
        try {
            for (SmartMandarinUserData.Unigram item : userData.unigrams(query)) {
                merged.put(item.text(), new Unigram(item.text(), item.probability(), item.backoff()));
            }
        } catch (RuntimeException ignored) {
            // Keep the cooked model available if the user database is busy.
        }
        String learned = learnedCandidate(query);
        if (learned != null && merged.containsKey(learned)) {
            Unigram item = merged.get(learned);
            merged.put(learned, new Unigram(learned, 0, item.backoff()));
        }
        ArrayList<Unigram> result = new ArrayList<>(merged.values());
        result.sort((left, right) -> Double.compare(right.probability(), left.probability()));
        return result;
    }

    private Double bigramProbability(String previousQuery, String currentQuery,
                                     String previousText, String currentText) {
        if (userData != null) {
            try {
                Double learned = userData.learnedBigram(previousQuery, currentQuery,
                        previousText, currentText);
                if (learned != null) return learned;
            } catch (RuntimeException ignored) {
                // Use the cooked bigram when the user database is busy.
            }
        }
        String query = previousQuery + " " + currentQuery;
        Map<BigramKey, Double> rows = bigramCache.get(query);
        if (rows == null) {
            rows = new HashMap<>();
            try (Cursor cursor = database.rawQuery(
                    "SELECT previous, current, probability FROM bigrams WHERE qstring = ?",
                    new String[]{query})) {
                while (cursor.moveToNext()) {
                    rows.put(new BigramKey(cursor.getString(0), cursor.getString(1)),
                            cursor.getDouble(2));
                }
            }
            bigramCache.put(query, rows);
        }
        return rows.get(new BigramKey(previousText, currentText));
    }

    @Override
    public void close() {
        database.close();
    }

    private String learnedCandidate(String query) {
        if (userData == null) return null;
        try {
            return userData.learnedCandidate(query);
        } catch (RuntimeException ignored) {
            return null;
        }
    }
}

File: Source/Loaders/iOS-Keyboard/KeyKeyEngine/Sources/KeyKeyEngine/SmartMandarinStore.swift
import Foundation

public struct SmartMandarinSegment: Equatable, Sendable {
    public let start: Int
    public let length: Int
    public let query: String
    public let text: String

    public init(start: Int, length: Int, query: String, text: String) {
        self.start = start
        self.length = length
        self.query = query
        self.text = text
    }
}

public struct SmartMandarinComposition: Equatable, Sendable {
    public let text: String
    public let segments: [SmartMandarinSegment]

    public init(text: String, segments: [SmartMandarinSegment]) {
        self.text = text
        self.segments = segments
    }
}

public struct SmartMandarinSelection: Equatable, Sendable {
    public let length: Int
    public let text: String

    public init(length: Int, text: String) {
        self.length = length
        self.text = text
    }
}

public struct SmartMandarinCandidate: Equatable, Sendable {
    public let length: Int
    public let text: String

    public init(length: Int, text: String) {
        self.length = length
        self.text = text
    }
}

public protocol SmartMandarinSource {
    func compose(readings: [String], overrides: [Int: String]) -> SmartMandarinComposition?
    func compose(
        readings: [String], selections: [Int: SmartMandarinSelection]
    ) -> SmartMandarinComposition?
    func candidates(
        for readings: [String], at index: Int, composition: SmartMandarinComposition?
    ) -> [String]
    func candidateOptions(
        for readings: [String], at index: Int, composition: SmartMandarinComposition?
    ) -> [SmartMandarinCandidate]
    func learnSelection(
        readings: [String], at index: Int, selected: String,
        composition: SmartMandarinComposition?
    )
    func learnSelection(
        readings: [String], at index: Int, candidate: SmartMandarinCandidate,
        composition: SmartMandarinComposition?
    )
    func learnConfirmedComposition(_ composition: SmartMandarinComposition)
}

public extension SmartMandarinSource {
    func compose(
        readings: [String], selections: [Int: SmartMandarinSelection]
    ) -> SmartMandarinComposition? {
        guard selections.values.allSatisfy({ $0.length == 1 }) else { return nil }
        return compose(readings: readings, overrides: selections.mapValues(\.text))
    }

    func candidateOptions(
        for readings: [String], at index: Int, composition: SmartMandarinComposition?
    ) -> [SmartMandarinCandidate] {
        candidates(for: readings, at: index, composition: composition).map {
            SmartMandarinCandidate(length: 1, text: $0)
        }
    }

    func learnSelection(
        readings: [String], at index: Int, selected: String,
        composition: SmartMandarinComposition?
    ) {}
    func learnSelection(
        readings: [String], at index: Int, candidate: SmartMandarinCandidate,
        composition: SmartMandarinComposition?
    ) {
        if candidate.length == 1 {
            learnSelection(
                readings: readings, at: index, selected: candidate.text,
                composition: composition
            )
        }
    }
    func learnConfirmedComposition(_ composition: SmartMandarinComposition) {}
}

/// A small, read-only Viterbi walker over the same `unigrams` and `bigrams`
/// tables used by macOS Manjusri. Queries stay lazy and cached so the keyboard
/// extension does not load the complete language model into its limited memory.
public final class SmartMandarinStore: SmartMandarinSource {
    private struct Unigram {
        let text: String
        let probability: Double
        let backoff: Double
    }

    private struct BigramKey: Hashable {
        let previous: String
        let current: String
    }

    private struct Path {
        let score: Double
        let lastProbability: Double
        let lastBackoff: Double
        let segments: [SmartMandarinSegment]
    }

    private let database: Database
    private let userData: SmartMandarinUserData?
    private let unigramStatement: Statement
    private let bigramStatement: Statement
    private var unigramCache: [String: [Unigram]] = [:]
    private var bigramCache: [String: [BigramKey: Double]] = [:]
    private let maximumSpan = 8

    public init(database: Database, userData: SmartMandarinUserData? = nil) throws {
        self.database = database
        self.userData = userData
        unigramStatement = try database.prepare(
            "SELECT current, probability, backoff FROM unigrams "
                + "WHERE qstring = ? ORDER BY probability DESC LIMIT 64"
        )
        bigramStatement = try database.prepare(
            "SELECT previous, current, probability FROM bigrams WHERE qstring = ?"
        )
    }

    public func compose(
        readings: [String], overrides: [Int: String] = [:]
    ) -> SmartMandarinComposition? {
        compose(readings: readings, selections: overrides.mapValues {
            SmartMandarinSelection(length: 1, text: $0)
        })
    }

    public func compose(
        readings: [String], selections: [Int: SmartMandarinSelection]
    ) -> SmartMandarinComposition? {
        guard !readings.isEmpty else {
            return SmartMandarinComposition(text: "", segments: [])
        }

        var paths = Array(repeating: [String: Path](), count: readings.count + 1)
        paths[0][""] = Path(score: 0, lastProbability: 0, lastBackoff: 0, segments: [])

        for start in readings.indices where !paths[start].isEmpty {
            let largestSpan = min(maximumSpan, readings.count - start)
            for length in 1...largestSpan {
                let end = start + length
                let query = readings[start..<end].joined()
                let overlapping = selections.filter { selectedStart, selection in
                    selectedStart < end && start < selectedStart + selection.length
                }
                if !overlapping.isEmpty && !(overlapping.count == 1
                    && overlapping[start]?.length == length) {
                    continue
                }

                var entries = unigrams(for: query)
                let learned = userData?.learnedCandidate(for: query)
                if let required = selections[start] {
                    entries = entries.filter { $0.text == required.text }
                }
                guard !entries.isEmpty else { continue }

                for entry in entries {
                    let segment = SmartMandarinSegment(
                        start: start, length: length, query: query, text: entry.text
                    )
                    for previousPath in paths[start].values {
                        let transition: Double
                        if let previous = previousPath.segments.last {
                            let fallback = previousPath.lastBackoff + entry.probability
                            let observed = bigramProbability(
                                previousQuery: previous.query,
                                currentQuery: query,
                                previousText: previous.text,
                                currentText: entry.text
                            )
                            transition = max(observed ?? fallback, fallback)
                        } else {
                            let observed = bigramProbability(
                                previousQuery: "!",
                                currentQuery: query,
                                previousText: "",
                                currentText: entry.text
                            )
                            transition = max(observed ?? entry.probability, entry.probability)
                        }

                        let learnedTransition = learned == entry.text
                            ? max(transition, 0) : transition
                        let path = Path(
                            score: previousPath.score + learnedTransition,
                            lastProbability: entry.probability,
                            lastBackoff: entry.backoff,
                            segments: previousPath.segments + [segment]
                        )
                        let stateKey = query + "\u{1f}" + entry.text
                        if path.score > (paths[end][stateKey]?.score ?? -.infinity) {
                            paths[end][stateKey] = path
                        }
                    }
                }
            }
        }

        // For an isolated syllable, match the desktop walker's first choice:
        // the end marker must not rerank the candidate shown to the user.
        guard let best = paths[readings.count].values.max(by: {
            let first = readings.count == 1 ? $0.score : finalScore($0)
            let second = readings.count == 1 ? $1.score : finalScore($1)
            return first < second
        }) else { return nil }
        return SmartMandarinComposition(
            text: best.segments.map(\.text).joined(), segments: best.segments
        )
    }

    public func candidates(
        for readings: [String], at index: Int, composition: SmartMandarinComposition?
    ) -> [String] {
        candidateOptions(for: readings, at: index, composition: composition).map(\.text)
    }

    public func candidateOptions(
        for readings: [String], at index: Int, composition: SmartMandarinComposition?
    ) -> [SmartMandarinCandidate] {
        guard readings.indices.contains(index) else { return [] }
        let previous = composition?.segments.last(where: { $0.start + $0.length == index })
        let previousBackoff = previous.flatMap { segment in
            unigrams(for: segment.query).first(where: { $0.text == segment.text })?.backoff
        } ?? 0
        var ranked: [(SmartMandarinCandidate, Double)] = []
        for length in 1...min(maximumSpan, readings.count - index) {
            let query = readings[index..<(index + length)].joined()
            let learned = userData?.learnedCandidate(for: query)
            for entry in unigrams(for: query) {
                let score: Double
                if learned == entry.text {
                    score = 0
                } else if let previous {
                    let fallback = previousBackoff + entry.probability
                    let observed = bigramProbability(
                        previousQuery: previous.query,
                        currentQuery: query,
                        previousText: previous.text,
                        currentText: entry.text
                    )
                    score = max(observed ?? fallback, fallback)
                } else {
                    let observed = bigramProbability(
                        previousQuery: "!", currentQuery: query,
                        previousText: "", currentText: entry.text
                    )
                    score = max(observed ?? entry.probability, entry.probability)
                }
                ranked.append((SmartMandarinCandidate(length: length, text: entry.text), score))
            }
        }
        ranked.sort { $0.1 > $1.1 }

        var seen = Set<String>()
        return ranked.compactMap { candidate, _ in
            seen.insert("\(candidate.length)\u{1f}\(candidate.text)").inserted
                ? candidate : nil
        }
    }

    public func learnSelection(
        readings: [String], at index: Int, selected: String,
        composition: SmartMandarinComposition?
    ) {
        learnSelection(
            readings: readings, at: index,
            candidate: SmartMandarinCandidate(length: 1, text: selected),
            composition: composition
        )
    }

    public func learnSelection(
        readings: [String], at index: Int, candidate: SmartMandarinCandidate,
        composition: SmartMandarinComposition?
    ) {
        guard readings.indices.contains(index), candidate.length > 0,
              index + candidate.length <= readings.count else { return }
        let previous = composition?.segments.last(where: { $0.start + $0.length == index })
        userData?.learnCandidate(
            query: readings[index..<(index + candidate.length)].joined(),
            current: candidate.text,
            previousQuery: previous?.query, previous: previous?.text
        )
    }

    public func learnConfirmedComposition(_ composition: SmartMandarinComposition) {
        userData?.learnComposition(composition)
    }

    private func finalScore(_ path: Path) -> Double {
        guard let last = path.segments.last else { return path.score }
        let observed = bigramProbability(
            previousQuery: last.query,
            currentQuery: "$",
            previousText: last.text,
            currentText: ""
        )
        return path.score + max(observed ?? path.lastBackoff, path.lastBackoff)
    }

    private func unigrams(for query: String) -> [Unigram] {
        let base: [Unigram]
        if let cached = unigramCache[query] {
            base = cached
        } else {
            base = unigramStatement.allRows(
                [query], columnCount: 3
            ).compactMap { row -> Unigram? in
                guard row.count == 3,
                      let probability = Double(row[1]), let backoff = Double(row[2]),
                      !row[0].isEmpty
                else { return nil }
                return Unigram(text: row[0], probability: probability, backoff: backoff)
            }
            unigramCache[query] = base
        }
        let learned = userData?.learnedCandidate(for: query)
        let custom = userData?.unigrams(for: query) ?? []
        var merged: [String: Unigram] = [:]
        for item in base { merged[item.text] = item }
        for item in custom {
            merged[item.text] = Unigram(text: item.text, probability: item.probability,
                                        backoff: item.backoff)
        }
        if let learned, let item = merged[learned] {
            merged[learned] = Unigram(text: learned, probability: 0,
                                      backoff: item.backoff)
        }
        return merged.values.sorted { $0.probability > $1.probability }
    }

    private func bigramProbability(
        previousQuery: String, currentQuery: String,
        previousText: String, currentText: String
    ) -> Double? {
        if let learned = userData?.learnedBigram(
            previousQuery: previousQuery, query: currentQuery,
            previous: previousText, current: currentText
        ) { return learned }
        let query = previousQuery + " " + currentQuery
        let rows: [BigramKey: Double]
        if let cached = bigramCache[query] {
            rows = cached
        } else {
            var loaded: [BigramKey: Double] = [:]
            for row in bigramStatement.allRows([query], columnCount: 3) where row.count == 3 {
                guard let probability = Double(row[2]) else { continue }
                loaded[BigramKey(previous: row[0], current: row[1])] = probability
            }
            bigramCache[query] = loaded
            rows = loaded
        }
        return rows[BigramKey(previous: previousText, current: currentText)]
    }
}
